src/transport/SecurePairingSession.h

```diff
@@ -0,0 +1,166 @@
+
+    /**
+     * @brief
+     *   Called when pairing fails with an error
+     *
+     * @param error error code
+     */
+    virtual void OnPairingError(CHIP_ERROR error) {}
+
+    /**
+     * @brief
+     *   Called when the pairing is complete and the new secure session has been established
+     *
+     */
+    virtual void OnPairingComplete() {}
+
+    virtual ~SecurePairingSessionDelegate() {}
+};
+
+class DLL_EXPORT SecurePairingSession
+{
+public:
+    SecurePairingSession(void);
+    SecurePairingSession(SecurePairingSession &&)      = default;
+    SecurePairingSession(const SecurePairingSession &) = default;
+    SecurePairingSession & operator=(const SecurePairingSession &) = default;
+    SecurePairingSession & operator=(SecurePairingSession &&) = default;
+
+    ~SecurePairingSession(void);
+
+    /**
+     * @brief
+     *   Initialize using setup PIN code and wait for pairing requests.
+     *
+     * @param mySetUpPINCode  Setup PIN code of the local device
+     * @param delegate        Callback object
+     *
+     * @return CHIP_ERROR     The result of initialization
+     */
+    CHIP_ERROR WaitForPairing(uint32_t mySetUpPINCode, uint32_t pbkdf2IterCount, const unsigned char * salt, size_t saltLen,
+                              SecurePairingSessionDelegate * delegate);
+
+    /**
+     * @brief
+     *   Create a pairing request using peer's setup PIN code.
+     *
+     * @param peerSetUpPINCode Setup PIN code of the peer device
+     * @param delegate         Callback object
+     *
+     * @return CHIP_ERROR      The result of initialization
+     */
+    CHIP_ERROR Pair(uint32_t peerSetUpPINCode, uint32_t pbkdf2IterCount, const unsigned char * salt, size_t saltLen,
+                    SecurePairingSessionDelegate * delegate);
+
+    /**
+     * @brief
+     *   Derive a secure session from the paired session. The API will return error
+     *   if called before pairing is established.
+     *
+     * @param info        Information string used for key derivation
+     * @param info_len    Length of info string
+     * @param session     Referene to the sescure session that will be
+     *                    initialized once pairing is complete
+     * @return CHIP_ERROR The result of session derivation
+     */
+    CHIP_ERROR DeriveSecureSession(const unsigned char * info, size_t info_len, SecureSession * session);
+
+    /**
+     * @brief
+     *   Handler for peer's messages, exchanged during pairing handshake.
+     *
+     * @param msg         Message sent by the peer
+     * @return CHIP_ERROR The result of message processing
+     */
+    CHIP_ERROR HandlePeerMessage(const MessageHeader & header, System::PacketBuffer * msg);
+
+private:
+    CHIP_ERROR Init(uint32_t setupCode, uint32_t pbkdf2IterCount, const unsigned char * salt, size_t saltLen,
+                    SecurePairingSessionDelegate * delegate);
+
+    CHIP_ERROR HandleCompute_pA(const MessageHeader & header, System::PacketBuffer * msg);
```

Comment:
Could you point to some doc in a comment explaining what pA/pB/cB/cA are? This requires deep protocol knowledge and people would likely want to read up on it.

Reply:
It's defined in spake2p spec here: https://www.ietf.org/id/draft-bar-cfrg-spake2plus-01.html
Also, you can find some information here: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/16/files?short_path=fb61aaf#diff-fb61aafa8ddbd3a30dfac9764c8dcc00

Reply:
Also added a comment in the code to point to the IETF spec.